utils/vector.hpp

```diff
@@ -0,0 +1,281 @@
+#pragma once
+#include <cstdint>
+
+namespace jbrush
+{
+template<class T>
+class Vector
+{
+public:
+    using reference = T&;
+    using iterator  = T*;
+    using const_reference = const T&;
+    using const_iterator  = const T*;
+    
+    /* Constructors */
+    Vector();
+    Vector(const Vector<T>&);
+    Vector(Vector<T>&&);
+    explicit Vector(uint32_t);
+    
+    /* Destructor */
+    ~Vector();
+    
+    /* Capacity */
+    uint32_t getSize() const;
+    uint32_t getCapacity() const;
+    void resize(uint32_t);
+    void resize(uint32_t, const T&);
+    void reserve(uint32_t);
+    
+    /* Member Functions */
+    void push_back(const T&);
+    void pop_back();
+    void clear();
+    
+    /* Operator Overloads */
+    Vector<T>& operator=(const Vector&);
+    Vector<T>& operator=(Vector&&);
+    reference  operator[](uint32_t);
+    
+    /* Iterators */
+    iterator begin();
+    iterator end();
+    const_iterator begin() const;
+    const_iterator end()   const;
+    void erase(uint32_t);
+    void insert(uint32_t, const T&);
+    
+private:
+    /* Helpers */
+    
+    /* Data */
+    T* array;
+    uint32_t size;
+    uint32_t capacity;
+};
+
+/* Constructors */
+template <class T>
+Vector<T>::Vector()
+    : size(0), capacity(16)
+{
+    array = new T[capacity];
+}
+
+template<class T>
+Vector<T>::Vector(const Vector<T>& vec)
+    : size(vec.size), capacity(vec.capacity)
+{
+    array = new T[vec.capacity];
+    for(unsigned int i = 0; i < vec.size; i++)
+    {
+        array[i] = vec.array[i];
+    }
+}
+
+template<class T>
+Vector<T>::Vector(Vector<T>&& vec)
+    : size(vec.size), capacity(vec.capacity), array(vec.array)
+{
+    vec.array = nullptr;
+}
+
+template<class T>
+Vector<T>::Vector(unsigned int size)
+    : size(size), array(nullptr)
```

Comment:
The size should not be touched. The capacity should be the one taking in the parameter.
```suggestion
    : size(0), capacity(cap), array(nullptr)
```